icicle/include/icicle/msm.h

```diff
@@ -0,0 +1,108 @@
+#pragma once
+
+#include <functional>
+
+#include "icicle/errors.h"
+#include "icicle/runtime.h"
+#include "icicle/utils/utils.h"
+#include "icicle/config_extension.h"
+
+#include "icicle/curves/affine.h"
+#include "icicle/curves/projective.h"
+#include "icicle/fields/field.h"
+
+namespace icicle {
+
+  /*************************** Frontend APIs ***************************/
+  /**
+   * @struct MSMConfig
+   * @brief Configuration for Multi-Scalar Multiplication (MSM).
+   */
+  struct MSMConfig {
+    icicleStreamHandle stream; /**< Stream for asynchronous execution. */
+    int precompute_factor;     /**< Number of extra points to pre-compute for each point. See the
+                                *   precompute_msm_bases function; precompute_factor passed there needs to be equal to the
+                                * one used here. Larger values decrease the number of computations to make, on-line memory
+                                * footprint, but increase the static memory footprint. Default value: 1 (i.e., don't
+                                * pre-compute). */
+    int c; /**< \f$ c \f$ value, or "window bitsize", which is the main parameter of the "bucket method" used to solve
+            * the MSM problem. Larger value means more on-line memory footprint but also more parallelism and less
+            * computational complexity (up to a certain point). Default value: 0 (the optimal value of \f$ c \f$ is
+            * chosen automatically). */
+    int bitsize;                /**< Number of bits of the largest scalar. Typically equals the bitsize of scalar field,
+                                 *   but if a different (better) upper bound is known, it should be reflected in this variable.
+                                 *   Default value: 0 (set to the bitsize of scalar field). */
+    int batch_size;             /**< Number of MSMs to compute. Default value: 1. */
+    bool are_bases_shared;      /**< Bases are shared for batch. Set to true if all MSMs use the same bases. Otherwise
+                                   expecting #bases==#scalars. Default value: true. */
```

Comment:
It's not clear at first that we mean the number of bases == number of scalars. The `#<name>` notation seems like it is a variable of some sort
```suggestion
    bool are_bases_shared;      /**< Bases are shared for batch. Set to true if all MSMs use the same bases. Otherwise, the number 
                                    of bases and number of scalars are expected to be equal. Default value: true. */
```